### ISSUE TITLE
Resource order matters for Create OME-Zarr (multiplexing), but isn't transparent 

### ISSUE THREAD

**Issue by jluethi** [fractal-analytics-platform/fractal-tasks-core#596]
As something for the discussion of resource refactor:
The order of resources needs to match the 0, 1, 2, 3 keys of the allowed_channels for Create OME-Zarr (multiplexing). But this is not transparent.
In the past, just adding them in the same order in the fractal-demos example then worked. With the refactor of dataset creation upon project creation, that seems to have changed slightly. Now, sometimes using this:

```
fractal dataset add-resource $PROJECT_ID $DS_IN_ID $INPUT_PATH/cycle1
fractal dataset add-resource $PROJECT_ID $DS_IN_ID $INPUT_PATH/cycle2
fractal dataset add-resource $PROJECT_ID $DS_IN_ID $INPUT_PATH/cycle3
```

Orders the resources as:
cycle1
cycle3
cycle2

and then the example fails.

-----

Not urgent to fix anything here. But learning for resource usage: It's already not very transparent and susceptible to such issues. Would argue in favour of moving such information directly into the actual task, instead of having it as part of the dataset.

**Comment by jluethi** [comment]
If one adds them in the correct order in the web interface, everything works as expected.
Trying to modify the resources of an existing input dataset to get them to the correct order does not work though, for some reason. Even removing all existing ones and adding new ones someone messes up the order again, I don't understand why.

For me, this argues in a direction of dropping input datasets and just making the paths to the images parameters of the create ome zarr tasks

**Comment by tcompa** [comment]
Thanks for reporting, this clearly looks like a bug (somewhere in fractal-web or fractal-server). We can/should fix it, but I also agree with:
> Would argue in favour of moving such information directly into the actual task, instead of having it as part of the dataset.

**Comment by jluethi** [comment]
In Fractal web, i get the correct order when adding it to a fresh dataset. Some funny things happen when I edit existing datasets.
From the CLI client, I don’t know why they aren’t ordered anymore in the condition they are sent. Maybe just whichever request finished first? E.g. if we’d add a 1s sleep between adding resources in the demo script, it may work as expected again?

**Comment by tcompa** [comment]
We debugged this a bit with @mfranzon

## Ordering in current DB is not specifically guaranteed

I confirm that fractal-server does not have specific logic to handle the order or the `Dataset.resource_list` relationship. This is different, e.g., from the case of `Workflow.task_list`, since that list is explicitly order according to a specific attribute (`WorkflowTask.order`).

This means that the most likely outcome (and the only one we could observe in our tests) is to have resources ordered by their DB ID. Some tests with fractal-web and fractal-client tend to show that this is the case. Independently on how we construct a set of resource, they will always be visible from a client ordered by their ID.

![image](https://github.com/fractal-analytics-platform/fractal-tasks-core/assets/3862206/1b018882-d631-4e38-9206-c196d380caf6)
![image](https://github.com/fractal-analytics-platform/fractal-tasks-core/assets/3862206/540c7ff8-e2df-467a-a127-8c425f3d8db7)

NOTE: this is in general an issue, since one should have a way of specifying the resource order without knowing IDs. It's not worth fixing this specific issue, however, since I think we all agree that this piece of information should be moved somewhere else (that is, to task input parameters).

## More information needed

@jluethi, we could not reproduce your error. 

After running
```
fractal dataset add-resource $PROJECT_ID $DS_IN_ID $INPUT_PATH/cycle1
fractal dataset add-resource $PROJECT_ID $DS_IN_ID $INPUT_PATH/cycle2
fractal dataset add-resource $PROJECT_ID $DS_IN_ID $INPUT_PATH/cycle3
```
we find that any client (web or CLI) shows the resources in the correct order.

Can you please report more details about where you observed this output:
> cycle1
> cycle3
> cycle2

Was it in the CLI/web client, or was it part of a task-error traceback? More info would be useful.

## What about the task?

Our understanding is that even if we provide a wrong `input_paths` (that is, with the wrong order) to the `create_ome_zarr_multiplex` task, it still runs through without errors. However, it is setting the wrong `acquisition` property (because we are reading it from the item order in the list), which is obviously leading to a wrong OME-Zarr.

Is this what happened in your case? Or did you actually observe a task error?

**Comment by tcompa** [comment]
More about: "what is the default ordering of `resource_list`?"
The answer is actually "it's an arbitrary ordering". The fact that we observed resources to be ordered according to their IDs, in the previous comment, is likely due to other external factors (e.g. using sqlite, or having tables of a certain size, or whatever changes the internal behavior of the SQL queries).
See e.g.
* https://stackoverflow.com/questions/44093038/why-is-the-order-of-sqlalchemy-instrumentedlist-not-persistent

TL;DR
We cannot rely on the resource order, in the current situation, even if it often tends to work as expected.

Question: are many many people are using the multiplexing task right now? If so, and if it works for them, it is mostly luck and/or due to external factors that we are not aware of.

**Comment by tcompa** [comment]
Something more:
tests performed as part of https://github.com/fractal-analytics-platform/fractal-server/pull/951 show a different behavior between postgres and sqlite, with sqlite sorting resources by ID and postgres by the order of insertions. The behavior is still undefined, but it's useful to be aware of differences in what would be observed.

**Comment by jluethi** [comment]
> we find that any client (web or CLI) shows the resources in the correct order.

I was running example 03 from fractal-demos. In the past, that always just worked. Now, it sometimes runs into the "wrong" order of datasets (i.e. not the order I then expect in the workflow).

> Is this what happened in your case? Or did you actually observe a task error?

Our test case contains different channels in different cycles. Thus, the allowed_channels list check fails when they are matched wrong (e.g. suddenly our second acquisition has more channels than it's allowed to have).
Also, having cycles in the wrong order afterwards can be problematic for downstream analysis.

My high-level take-away is that we want to rethink this about whether those input paths should just be a task argument => makes it explicit in one location only.

> Question: are many many people are using the multiplexing task right now? If so, and if it works for them, it is mostly luck and/or due to external factors that we are not aware of.

We don't have too many multiplexing users yet, only some early tests. If they add their datasets via the web interface (which I think everyone does), they still ended up ordered correctly for me always. But the new explicit ordering by ID is very helpful to ensure that!
Even when someone runs from a script, if they have the same channels in all cycles (often the case), it wouldn't actually fail, just produce a slightly unexpected arrangement of acquisitions.

**Comment by tcompa** [comment]
Thanks for the additional details.

Next fractal-server version (as of [this PR](https://github.com/fractal-analytics-platform/fractal-server/pull/951)) will at least have a predictable by-ID ordering, which will work as expected in most cases. It's clearly not *the* solution, and I agree with you on this:

> My high-level take-away is that we want to rethink this about whether those input paths should just be a task argument => makes it explicit in one location only.

**Comment by jluethi** [comment]
Yes, saw that PR. Great to have this more predictable for now, let's see what direction we go with it :)

**Comment by tcompa** [comment]
In Fractal V2 there are "resources" any more, and the list of image folders is part of the task arguments (which removes any list-order ambiguity).

Closing.
